### DISCUSSION TITLE
A Rotated Tiles Proof of Concept

### DISCUSSION THREAD

**Discussion by ProjectMobius13** [TrianglyRU/OrbinautFramework#41] [Show and tell]
> Note: Amended with information that was mistakenly omitted before.

A proof of concept showing collision tiles rotated in the room editor working after a little tweaking. Full disclosure: there may be some edge cases I haven't found that result in some wonky behavior. It's also not the cleanest code, and there is likely a more elegant solution. But it works!
![OrbinautFramework2-GameMaker-Editor](https://github.com/user-attachments/assets/91a33ca2-d25a-4e7b-b14d-5f909e3f2bd8)
![OrbinautFramework2-Gameplay](https://github.com/user-attachments/assets/edcb6a8f-bba5-48cd-8873-e2d2350b7a4b)

Here are the changes to the scripts needed to make this work:

```gml
function tile_get_width(_tiledata, _y)
{
	var _index = tile_get_index(_tiledata) % ENGINE_TILE_COUNT;
	
	if _index <= 0
	{
		return 0;
	}
	
	// Calculate width index
	var _width_index;
	
	var _tile_mode = tile_get_rotate(_tiledata) + (10*tile_get_mirror(_tiledata)) + (100*tile_get_flip(_tiledata));
	switch (_tile_mode)
	{
		/*
			Bits: [ Flip | Mirror | Rotate ]
		*/
		case 000: //Normal
		case 010: //Mirrored Horizontally: Height = Reverse Height
		case 001: //Rotated: Width = Height; Height = Reversed Width
		case 101: //Flipped and Rotated: Width = Height; Height = Width
			_width_index = _y % ENGINE_TILE_SIZE;
		break;
		
		case 100: //Flipped vertically: Width = Reverse Width
		case 110: //Flipped and Mirrored: Width = Reverse Width; Height = Reverse Height
		case 011: //Mirrored and Rotated: Width = Reversed Height; Height = Reversed Width
		case 111: //Flipped, Mirrored, and Rotated: Width = Reversed Height; Height = Width;
			_width_index = ENGINE_TILE_SIZE - 1 - _y % ENGINE_TILE_SIZE;
		break;
	}
	
	if tile_get_rotate(_tiledata)
	{
		return c_framework.collision.tile_height[_index][_width_index];
	}
	else
	{
		return c_framework.collision.tile_width[_index][_width_index];
	}
}
```
```gml
function tile_get_height(_tiledata, _x)
{
	var _index = tile_get_index(_tiledata) % ENGINE_TILE_COUNT;
	
	if _index <= 0
	{
		return 0;
	}
	
	// Calculate height index
	var _height_index;
	
	var _tile_mode = tile_get_rotate(_tiledata) + (10*tile_get_mirror(_tiledata)) + (100*tile_get_flip(_tiledata));
	switch (_tile_mode)
	{
		/*
			Bits: [ Flip | Mirror | Rotate ]
		*/
		case 000: //Normal
		case 100: //Flipped Vertically
		case 101: //Flipped and Rotated: Width = Height; Height = Width
		case 111: //Flipped, Mirrored, and Rotated: Width = Reversed Height; Height = Width;
			_height_index = _x % ENGINE_TILE_SIZE;
		break;
			
		case 010: //Mirrored Horizontally: Height = Reverse Height
		case 110: //Flipped and Mirrored: Height = Reverse Height
		case 001: //Rotated: Width = Height; Height = Reversed Width
		case 011: //Mirrored and Rotated: Width = Reversed Height; Height = Reversed Width
			_height_index = ENGINE_TILE_SIZE - 1 - _x % ENGINE_TILE_SIZE;
		break;
	}
	
	if tile_get_rotate(_tiledata)
	{
		return c_framework.collision.tile_width[_index][_height_index];
	}
	else
	{
		return c_framework.collision.tile_height[_index][_height_index];
	}
}
```

In both **_tile_find_h_** and **_tile_find_v_**, you will find this block of code:
```gml
	// Calculate final distance and angle
	var _flip = tile_get_flip(_tiledata);
	var _mirror = tile_get_mirror(_tiledata);
	var _ang = tile_get_angle(_tiledata);
```

Replace it with:
```gml
	// Calculate final distance and angle
	var _flip = tile_get_flip(_tiledata);
	var _mirror = tile_get_mirror(_tiledata);
	var _ang = tile_get_angle(_tiledata);
	var _rot = tile_get_rotate(_tiledata);
```

Just below that in both **_tile_find_h_** and **_tile_find_v_**, you will find this block of code:
```gml
if _flip
{
	_ang = (540 - _ang) % 360;
}
if _mirror
{
	_ang = 360 - _ang;
}
```

Replace it with:
```gml
if _flip
{
	_ang = (540 - _ang) % 360;
}
if _mirror
{
	_ang = 360 - _ang;
}
if _rot
{
	_ang = (360 + _ang - 90) % 360;
}
```

And voilà! No more breaking when tiles are rotated in the room editor.

**Comment by TrianglyRU** [comment]
Oh wow, good stuff! Really clever usage of a bitmask

**Reply by ProjectMobius13** [reply]
Thanks!

**Reply by Otukuoi** [reply]
Where would I copy>paste the first two? I put them in tile_get_width and tile_get_height respectively & the project won't even load without crashing.

**Reply by ProjectMobius13** [reply]
> Where would I copy>paste the first two? I put them in tile_get_width and tile_get_height respectively & the project won't even load without crashing.

The first two are the entire functions. So all text in the tile_get_width and tile_get_height should be replaced with those.

**Reply by Otukuoi** [reply]
I did that prior to asking this, and upon touching a collision tile the game instantly crashes. I retried it to see if maybe I didn't copy all of it, or erase all the original code. Is there something else that has been modified that isn't shared here? Or do you have a clue as to what I did wrong?

**Reply by ProjectMobius13** [reply]
Without seeing the crash message, I'm afraid I don't have much hope of helping you fix your problem. The instructions here was all I modified in my copy of Orbinaut 2.

**Comment by Otukuoi** [comment]
I may provide the crash messages soon, but I ended up just replacing it with the original code. If I implement it again soon I'll be sure to send the crash messages (if it still doesn't work)

**Reply by ProjectMobius13** [reply]
Sounds good!

**Reply by TrianglyRU** [reply]
Decided to check it out myself, the game does not crash but player basically cannot properly land on solid ground. There is probably something else you forgot to mention in the guide

**Reply by Otukuoi** [reply]
For me, it'd just crash the second ground was touched, no matter what type of tile it was. Maybe it's the collision I was using? (S1) 

In any case, it does not work so there's likely something else they forgot. Perhaps a very small modification to another script to make it work.

**Reply by ProjectMobius13** [reply]
Hmm.... I'll check it out once I'm home. If I find something I forgot, I'll edit the post

**Reply by ProjectMobius13** [reply]
There was a mistake in the original guide. The guide has been corrected and verified on my PC on a fresh project. Let me know if you run into any more troubles!

**Reply by Otukuoi** [reply]
Alright, thanks a lot!
